docs: add migration guide to help users move from Proxy to Edge (#179)
Unleash/unleash-edge: migration-guide.md

# Migrating from the Unleash Proxy to Edge

> Since Edge is built to be a near drop in replacement for the Unleash Proxy, this guide should take you through the differences between Edge and the Unleash Proxy, and how to migrate from the Unleash Proxy to Edge, and, in cases where the feature set isn't equivalent, how to use achieve the same results in Edge. A full docker compose is provided [here](./examples/docker-compose.yml), that will spin up Edge, Unleash and Redis, to allow you to understand the configuration options in context. After starting the compose, you should be able to access the Unleash UI at `http://localhost:4242`, add a toggle and cURL the edge instance with:

Reviewer: ```suggestion
Since Edge is built to be a near drop in replacement for the Unleash Proxy, this guide should take you through the differences between Edge and the Unleash Proxy, and how to migrate from the Unleash Proxy to Edge, and, in cases where the feature set isn't equivalent, how to achieve the same results in Edge. A full docker compose is provided [here](./examples/docker-compose.yml), that will spin up Edge, Unleash and Redis, to allow you to understand the configuration options in context. After starting the compose, you should be able to access the Unleash UI at `http://localhost:4242`, add a toggle and cURL the edge instance with:
```

``` sh
curl --location --request GET 'http://0.0.0.0:3063/api/client/features' \
--header 'Content-Type: application/json' \
--header 'Authorization: default:development.unleash-insecure-api-token' \
--data-raw ''
```

## Not supported

> - Custom Strategies are not supported in Edge today. If you need to use a custom strategy, you will need to use the Unleash Proxy. We believe that the overwhelming majority of custom strategies are better expressed as a set of Unleash Constraints. If you have a situation where this absolutely isn't the case, please raise this as an issue with details on what you're trying to achieve. We are looking into supporting custom strategies in Edge in the future.

Reviewer 2: Can we add some doc links here? Also, because each bullet has a lot of text, it could maybe be good to instead break it up into subsections? But again, maybe not? I'll let you pick which one.

```suggestion
- [Custom Strategies](https://docs.getunleash.io/reference/custom-activation-strategies) are not supported in Edge today. If you need to use a custom strategy, you will need to use the Unleash proxy. We believe that the overwhelming majority of custom strategies are better expressed as a set of [strategy constraints](https://docs.getunleash.io/reference/strategy-constraints). If you have a situation where constraints **cannot** replace your strategy, please raise this as an issue with details on what you're trying to achieve. We are looking into supporting custom strategies in Edge in the future.
```

- Legacy Proxy Tokens. If you're using the Proxy, you may be using the legacy Proxy token format. These are not supported in Edge. You will need to create a new front end SDK token in the Unleash UI and use that. These are the same tokens that the front end API requires. Because of the way Edge handles API tokens, this is not a feature we're planning to support.

- CORS configuration. Edge does not support CORS configuration, we accept any request. If you need this feature, please raise an issue or submit a PR.

- Context Enrichers. The Unleash Proxy provides an experimental option to automatically enrich requests with additional context. This is not supported in currently in Edge, we're open to adding this feature to Edge, opening an issue would help us to know that this is a valuable feature to you.

> - JS integration. Edge is written in Rust, rather than JavaScript, like the Proxy. If you're wrapping the Proxy code in your own JavaScript, you'll need to port that equivalent code to Rust. Generally, if you have a deep integration on the existing Proxy, we would recommend not moving to Edge or to reaching out to us to discuss your use case.

Reviewer 2: The first sentence was a little hard to parse. Could we do something like this? Also the last sentence could use some clarification, because right now it kind of reads like "we recommend not reaching out to us", but I think you meant the opposite?

```suggestion
- JS integration. Edge is written in Rust. The Unleash proxy is written JavaScript. If you're wrapping the Proxy code in your own JavaScript, you'll need to port that equivalent code to Rust. Generally, if you have a deep integration on the existing Proxy, we would recommend either not moving to Edge, or to reach out to us to discuss your use case.
```

Author: Oh my word, that's hillarious, the result of too many edits. I definitely did not mean to tell people not to reach out to us

> ## Existing features in new ways

Reviewer 2: Definitely not a blocker, but this heading is a little confusing: can it be made clearer?

Author: Yuh, probably. So this is to capture features that have changed major ways - enough to make them not a one to one mapping. How about "Significant Changes to Proxy Features"?

Reviewer 2: Yeah, I like that!

    ``` sh
    ./edge edge --upstream-url http://localhost:4242 --redis-url redis://localhost:6379

    ```

- Multiple environments. The Unleash proxy only supports a single environment, provided by your API key when starting up. Edge does not suffer from this limitation, any connecting SDK can freely use API tokens to scope its request to an environment, project or both, so long as the upstream Unleash instance has that token. This is dynamically determined by Edge, so adding a new token to the upstream Unleash and using that token in an SDK talking to Edge will work without restarting Edge or changing configuration.

> - Scaling out Edge. Edge is built with high performance scaling as a first class citizen. Typically, the Proxy would be scaled out by spinning up multiple instances, due to the limitations in JS engines that prevent using all available cores. The other reason for potentially scaling out the Proxy is that the Proxy doesn't support multiple environments. Edge doesn't have these requirements, multiple environments are supported out the box and Edge, by default, will use all the processing power that's available to it to scale up. Edge runs cold, unless you're running Edge at extreme scaling levels, it's very likely that you won't need to scale out your Edge instance at all, in fact, its more likely that you want to reduce the resources available to Edge.

Reviewer 2: Should that be a full stop?
```suggestion
- Scaling out Edge. Edge is built with high performance scaling as a first class citizen. Typically, the Proxy would be scaled out by spinning up multiple instances, due to the limitations in JS engines that prevent using all available cores. The other reason for potentially scaling out the Proxy is that the Proxy doesn't support multiple environments. Edge doesn't have these requirements, multiple environments are supported out the box and Edge, by default, will use all the processing power that's available to it to scale up. Edge runs cold. Unless you're running Edge at extreme scaling levels, it's very likely that you won't need to scale out your Edge instance at all; in fact, its more likely that you want to _reduce_ the resources available to Edge.
```

Author: It should!

> - Scaling out Edge. Edge is built with high performance scaling as a first class citizen. Typically, the Proxy would be scaled out by spinning up multiple instances, due to the limitations in JS engines that prevent using all available cores. The other reason for potentially scaling out the Proxy is that the Proxy doesn't support multiple environments. Edge doesn't have these requirements, multiple environments are supported out the box and Edge, by default, will use all the processing power that's available to it to scale up. Edge runs cold, it uses Unless you're running Edge at extreme scaling levels, it's very likely that you won't need to scale out your Edge instance at all, in fact, its more likely that you want to reduce the resources available to Edge.

Reviewer: ```suggestion
- Scaling out Edge. Edge is built with high performance scaling as a first class citizen. Typically, the Proxy would be scaled out by spinning up multiple instances, due to the limitations in JS engines that prevent using all available cores. The other reason for potentially scaling out the Proxy is that the Proxy doesn't support multiple environments. Edge doesn't have these requirements, multiple environments are supported out the box and Edge, by default, will use all the processing power that's available to it to scale up. Edge runs cold, unless you're running Edge at extreme scaling levels, it's very likely that you won't need to scale out your Edge instance at all, in fact, its more likely that you want to reduce the resources available to Edge.
```

## Existing features with new configuration options

This section unpacks the small changes in Edge from the Proxy. These are ports of existing features or configuration that have small changes. These shouldn't affect how you use Edge and the ideas here are similar to the Proxy, only small details have changed.

> - Unleash URL. The Proxy requires that you specify an Unleash URL to the upstream server, in the format https://{unleashUrl}/api. Edge has changed this, the URL that Edge requires is https://{unleashUrl}, without the `/api` suffix.

Reviewer 2: This is great! 🏆

- Backend SDK support. The Proxy does support connecting to backend SDKs, but it requires some configuration and setting some experimental feature flags. Edge supports this out the box, no additional configuration is required.

## New features

- Offline mode. Edge can be used in a purely offline mode. While the Proxy does support this, it requires a little work to get there. Edge, on the other hand, supports this directly and out the box. This can be useful for testing or development environments where you don't want to have to run an Unleash instance or want to freeze your configuration in place. Example run command:

    ``` sh
    ./edge offline --tokens "*.development.some-test-token"  --bootstrap-file ./examples/features.json
    ```

>     This will start Edge in offline mode and set the initial feature set to the contents of the `features.json` file. Edge will not send metrics to upstream Unleash instances, update its feature information or dynamically resolve tokens. Note that you must set a token or tokens on startup in this mode - Edge will only use this set of tokens to validate incoming requests, this doesn't have to be a valid Unleash token, so this is very similar to the original Proxy tokens.

Reviewer 2: ```suggestion
    This will start Edge in offline mode and set the initial feature set to the contents of the `features.json` file. Edge will not send metrics to upstream Unleash instances, update its feature information, or dynamically resolve tokens. Note that you must set a token or tokens on startup in this mode - Edge will only use this set of tokens to validate incoming requests, this doesn't have to be a valid Unleash token, so this is very similar to the original Proxy tokens.
```

- Daisy chaining. Edge is capable of both mimicking an Unleash server and talking to an Unleash server. This means that Edge can use another Edge instance as another upstream source. This can be extremely useful for high scaling scenarios where you can have a single instance close to your SDKs that retrieves all toggles for all projects and environments and specific Edge instances syncing to the parent Edge instance, with much more narrowly defined API tokens for security reasons. No extra configuration is needed to achieve this, simply point the downstream Edge instance at the upstream Edge instance and it will automatically resolve the tokens and features from the upstream Edge instance. Metrics will propagate up the daisy chain from downstream Edge instances to upstream instances until the metrics hit the end of the chain in either an offline Edge instance or an upstream Unleash instance. Note that if the daisy chain terminates in an offline Edge instance, the metrics will be discarded.

> 
> 

Reviewer 2: Silly thing, but do we need three blank lines at the end of the file? Isn't one enough?

```suggestion

```

Author: It's more than enough, this is just me apparently being over eager with the enter key

> - Daisy chaining. Edge is capable of both mimicking an Unleash server and talking to an Unleash server. This means that Edge can use another Edge instance as another upstream source. This can be extremely useful for high scaling scenarios where you can have a single instance close to your SDKs that retrieves all toggles for all projects and environments and specific Edge instances syncing to the parent Edge instance, with much more narrowly defined API tokens for security reasons. No extra configuration is needed to achieve this, simply point the downstream Edge instance at the upstream Edge instance and it will automatically resolve the tokens and features from the upstream Edge instance. Metrics will propagate up the daisy chain from downstream Edge instances to upstream instances until the metrics hit the end of the chain in either an offline Edge instance or an upstream Unleash instance. Note that if the daisy chain terminates in an offline Edge instance, the metrics will be discarded.

Reviewer: PRobably also worth mentioning that if the daisy chain ends in an offline Edge instance, it must have been started with all tokens used throughout the chain.

Author: Yup, good point, will add!
